### PR TITLE
[For 2.6.3] Fix/2811  fix reference header fallback

### PR DESCRIPTION
**Pull request by inkhey**
related to #2811
Fix References Headers,
This new code allow reply to email with one email address notification on mail client that properly follow the email rfc and feed references header properly like thunderbird and roundcube.
 
## Checkpoints

The goal of this section is to help code reviewer and merge request author to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

**For code reviewer**

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

**For developper**

- [x] Manual tests done to ensure stability on whole application layers, issue expectation follow
- [x] Original issue have been updated with a short resume of implemented solution

**For quality team**

- [x] Tested by quality team

### CODE REVIEW

File: backend/doc/setting.md [hunk 1]
```diff
@@ -340,6 +340,48 @@ Configure how to handle invitation of non-existent users in Tracim with these pa
...
+behavior, and does break reply to feature too .
+Email from these client are retrieved but can't be correctly associated to one item.
+
+You can do this with a configuration like that:
+
+    email.notification.from.email = tracim.notification@supersmtpserver.ndd
+    email.notification.reply_to.email = tracim.notification@supersmtpserver.ndd
+    # References headers doesn't have to be a true email address.
+    email.notification.references.email = tracim.content.{content_id}@supersmtpserver.ndd
+
```
Comment (PhilippeAccorsi) [inline backend/doc/setting.md:382]: I not really understand here if its possible to reply with this email configuration.

Comment (inkhey) [reply backend/doc/setting.md:382]: Yes, it does work on some client but does not on some others. It depends on how the client react.
It should normally work as the normal behavior is to resend references, that's why it work on roundcube and thunderbird, but not with sogo (the email is received but email fetcher can't associate the email to a specific content).

**Comment by inkhey** [comment]
will need merge fix with #2687

File: backend/doc/setting.md (hunk 1 above)
Comment (PhilippeAccorsi) [reply backend/doc/setting.md:382]: ok.

**Comment by PhilippeAccorsi** [comment]
What kind of test can I make to check if this fix working ?

File: backend/tracim_backend/lib/mail_fetcher/email_fetcher.py [hunk 2]
```diff
@@ -171,6 +170,8 @@ def find_key_from_mail_address(
                 cls, "pattern {} does not match email address {} ".format(pattern, mail_address)
             )
             return None
+        else:
+            return None
```
Comment (raphj) [inline backend/tracim_backend/lib/mail_fetcher/email_fetcher.py:174]: replace these three lines by a

```python
    return None
```

outside the if statement

File: backend/tracim_backend/lib/mail_notifier/notifier.py [hunk 3]
```diff
@@ -300,7 +299,13 @@ def notify_content_update(self, event_actor_id: int, event_content_id: int) -> N
             # contains only message_id from parents post in thread.
             # To link this email to a content we create a virtual parent
             # in reference who contain the content_id.
-            message["References"] = formataddr(("", reference_addr))
+            references = formataddr(("", reference_addr))
```
Comment (raphj) [inline backend/tracim_backend/lib/mail_notifier/notifier.py:302]: You probably can remove the use of `formataddr`. From [its documentation ](https://docs.python.org/3/library/email.utils.html#email.utils.formataddr): "If the first element of pair is false, then the second element is returned unmodified."

File: backend/tracim_backend/lib/mail_notifier/notifier.py [hunk 4]
```diff
@@ -300,7 +299,13 @@ def notify_content_update(self, event_actor_id: int, event_content_id: int) -> N
...
+            # INFO - G.M - 2020-04-03 - Enforce bracket in references header if not provided,
+            # formataddr add them only if label is not empty, but we need them here to ensure best
+            # compat from parsing software
+            if reference_addr[0] != "<":
```
Comment (raphj) [inline backend/tracim_backend/lib/mail_notifier/notifier.py:306]: Shoud `reference_addr` be checked for emptiness here, to avoid an IndexError?

Comment (inkhey) [reply backend/tracim_backend/lib/mail_notifier/notifier.py:306]: good suggestion.

File: backend/tracim_backend/lib/mail_notifier/notifier.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/lib/mail_notifier/notifier.py]: modified in EmailAdress.adress property (after pulling develop code).

File: backend/tracim_backend/lib/mail_notifier/notifier.py (hunk 4 above)
Comment (inkhey) [reply backend/tracim_backend/lib/mail_notifier/notifier.py]: done in EmailAddress.address

File: backend/tracim_backend/lib/mail_fetcher/email_fetcher.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/lib/mail_fetcher/email_fetcher.py]: done.

**Comment by inkhey** [comment]
@PhilippeAccorsi To test this, set email reply configuration with "One email address notification" and test with thunderbird or roundcube if behavior is ok.

**Comment by PhilippeAccorsi** [comment]
Ok, thanks.

**Comment by PhilippeAccorsi** [comment]
Tested and verified on branch. For testing I used this configuration (available in documentation):

~~~
email.notification.from.email = xxx.xxx@xxx.fr
email.notification.reply_to.email = xxx.xxx@xxx.fr
email.notification.references.email = xxx.xxx.{content_id}@xxx.fr
~~~

**Comment by PhilippeAccorsi** [comment]
I just wait one reply about typo and we can merge.

**Comment by PhilippeAccorsi** [comment]
> I just wait one reply about typo and we can merge.

I have tested `email.notification.references.email = xxx.xxx.{content_id}@xxx.fr` also with `email.notification.references.email = xxx.xxx+{content_id}@xxx.fr` and it is working good.

File: backend/doc/setting.md [hunk 5]
```diff
@@ -340,6 +340,48 @@ Configure how to handle invitation of non-existent users in Tracim with these pa
...
+    email.notification.reply_to.email = tracim.content.{content_id}@supersmtpserver.ndd
+    email.notification.references.email = tracim.content.{content_id}@supersmtpserver.ndd
+
+#### One email address notification
+
+if you can't provide a wildcard and can't support subadressing and want a reply-to feature with only one named email address,
+as a last resort, you can set the References Header.
+
+However, be aware that support across email clients is not great.
+clients. Some email client (like Sogo) do not properly return the References Header in mail replies, which breaks threading
```
Comment (raphj) [inline backend/doc/setting.md:372]:
```suggestion
Some email client (like Sogo) do not properly return the References Header in mail replies, which breaks threading
```

**Pull request merged**